Less wrapper elements

flash1293 opened the pull request:
Based on https://github.com/chandlerprall/eui/pull/17

This PR reduces the number of wrapper elements per cell - as long as the cell is not hovered, there's just the regular cell element with one wrapper plus a hidden paragraph for screen reader a11y. This one last wrapper element is still in place because the CSS got pretty ugly otherwise and it didn't seem to have too much influence on performance.

This also prevents a double-render by waiting for the header row to get a height first before placing the data cells.

This also fixes a bug with `onMouseLeave` which is not reliable, this PR solves it via css to only show the buttons in the cell is actually hovered.

flash1293 committed:
only show buttons and popover if necessary

flash1293 committed:
improve

flash1293 committed:
make sure buttons and popover are only rendered if necessary

flash1293 committed:
Merge branch 'feature/virtualized-datagrid' into flash1293/datagrid-no-popover

flash1293 committed:
clean up

flash1293 committed:
reduce rerenders

flash1293 committed:
remove console logs

flash1293 committed:
fix tests

flash1293 committed:
convert to single mutation observer

flash1293 committed:
Merge remote-tracking branch 'chandlerprall/feature/virtualized-datagrid' into flash1293/less-mutation-observers

flash1293 committed:
reduce number of wrapper elements if not necessary

flash1293 committed:
fix some bugs

chandlerprall commented on src/components/datagrid/data_grid.test.tsx:
```diff
@@ -2041,9 +2050,21 @@ describe('EuiDataGrid', () => {
+      expect(findTestSubject(component, 'alertAction').exists()).toBe(false);
+      expect(findTestSubject(component, 'happyAction').exists()).toBe(false);
+
+      await act(async () => {
```
Wrapping this interaction with `act` is not necessary.

chandlerprall commented on src/components/datagrid/data_grid_cell.tsx:
```diff
@@ -98,8 +98,10 @@ export interface EuiDataGridCellProps {
 interface EuiDataGridCellState {
   cellProps: CommonProps & HTMLAttributes<HTMLDivElement>;
   popoverIsOpen: boolean; // is expansion popover open
+  renderPopoverOpen: boolean; // wait one render cycle to actuall render the popover as open
```
```suggestion
  renderPopoverOpen: boolean; // wait one render cycle to actually render the popover as open
```

chandlerprall commented on src/components/datagrid/data_grid_cell.tsx:
```diff
@@ -310,6 +321,14 @@ export class EuiDataGridCell extends Component<
   };
 
+  closePopover = () => {
+    this.setState({ popoverIsOpen: false }, () =>
```
This nested setState doesn't appear necessary for closing the popover

flash1293 replied:
I thought so as well originally, but it has the same problem (not always properly unmounted, leading to hanging popovers). I added a comment

chandlerprall commented on src/components/datagrid/data_grid_body.tsx:
```diff
@@ -523,19 +529,41 @@ export const EuiDataGridBody: FunctionComponent<EuiDataGridBodyProps> = (
 
-    if (boundingRect.height !== unconstrainedHeight) {
+    if (
+      boundingRect.height !== unconstrainedHeight &&
+      height !== boundingRect.height
+    ) {
       setHeight(boundingRect.height);
     }
-    if (boundingRect.width !== unconstrainedWidth) {
+    if (
+      boundingRect.width !== unconstrainedWidth &&
+      width !== boundingRect.width
+    ) {
```
This change introduces a dependency on `height` & `width` to the useEffect, which would need to be ignored. The previous version should function identically, as `setState` is effectively a no-op if called with its present value.

flash1293 replied:
I added this because it's a quite costly no-op as the grid is re-rendering as a result of the state setting (even if the value itself isn't changing). Another way would be to move this part https://github.com/chandlerprall/eui/pull/18/files#diff-7fd629564b805d7d5c25b7c9bc1e570afc0d443fa880192698fc13fe9776a729R588-R623 into a memoized component, but I didn't like the resulting component because a lot of props have to be passed down.

chandlerprall replied:
The noop includes not triggering another render - https://codesandbox.io/s/laughing-cohen-suyii?file=/src/App.js

https://user-images.githubusercontent.com/313125/105218891-fdc4bc80-5b12-11eb-8a36-70271bcc9ccf.mov

flash1293 replied:
Hah, TIL. I was under the impression I tested this and it re-rendered, guess I confused it with some other change I had going on. My bad, I will adjust the PR.

chandlerprall replied:
No worries, I had to triple check my statements :)

chandlerprall commented on src/components/datagrid/data_grid_cell.tsx:
```diff
@@ -39,13 +39,13 @@ import {
 import { EuiDataGridCellButtons } from './data_grid_cell_buttons';
 import { EuiDataGridCellPopover } from './data_grid_cell_popover';
 
+let count = 0;
```
This and the related console.log are probably not necessary :)

flash1293 replied:
I thought they were a nice touch :)

chandlerprall commented on src/components/datagrid/data_grid_cell.tsx:
```diff
@@ -243,14 +248,20 @@ export class EuiDataGridCell extends Component<
       if (!this.props.style) return true;
-      if (nextProps.style.top !== this.props.style.top) return true;
+      if (nextProps.style.top !== this.props.style.top) {
+        console.log(`${nextProps.style.top} from ${this.props.style.top}`);
```
remove console.log

chandlerprall commented on src/components/datagrid/data_grid_cell.tsx:
```diff
@@ -352,7 +378,11 @@ export class EuiDataGridCell extends Component<
           case keys.F2:
             event.preventDefault();
-            this.setState({ popoverIsOpen: true });
+            this.setState({ popoverIsOpen: true }, () =>
```
Please add a comment here describing why the double-setState is necessary.

chandlerprall reviewed (changes requested):
Wow, this is impressively faster. Thank you for looking into all of this!

#### Cell action should appear immediately on click

One thing we decided with cell actions is that they should appear immediately onFocus instead of being delayed like the popover button

#### Icon animations reset when expanding the popover

When clicking on the popover button, the cell actions hide and then re-animate in, they should stay present. This may be fixed for free with the previous item.

https://user-images.githubusercontent.com/313125/105093841-94847100-5a60-11eb-843d-66ee548fdc62.mov

chandlerprall commented:
Quick testing pass on the _Data grid focus_ page:

* "When removing focus from the grid and then returning, the last focused cell remains focused." is no longer working
* "Clicking on an interactive element within a cell the focus should always remain on that element, not shift to the cell or another element unless a subsequent user action changes it." does not work if the column is expandable
* Not discussed in a rule, but the popover for the expandable + 2 interactives column doesn't have a focus trap.

flash1293 committed:
resolve some issues

flash1293 committed:
resolve some issues

flash1293 committed:
resolve some issues

flash1293 commented:
Thanks for the detailed look @chandlerprall 

I addressed some issues, I'm not sure about these two:

> "Clicking on an interactive element within a cell the focus should always remain on that element, not shift to the cell or another element unless a subsequent user action changes it." does not work if the column is expandable
> Not discussed in a rule, but the popover for the expandable + 2 interactives column doesn't have a focus trap.

Can we do an offline sync about those?

flash1293 committed:
fix leftover problems

flash1293 commented:
@chandlerprall  Fixed the problems discussed offline

chandlerprall commented:
Additional changes look good, replied to the setState/noop thread.

snide committed:
fix prettier / ts issues

snide commented:
@flash1293 Here's a branch of a branch of a branch with some small fixes to this PR.

https://github.com/flash1293/eui/pull/1

flash1293 committed:
Merge pull request #1 from snide/cleanup/less-wrappers

fix prettier / ts issues

flash1293 committed:
remove unnecessary checks

flash1293 committed:
remove unnecessary parameter

flash1293 commented:
@chandlerprall Cleaned up remaining stuff, I think it's worth another look

chandlerprall reviewed (approved):
thanks again @flash1293 !!
